### PR TITLE
Support unhashable list of target.json

### PR DESCRIPTION
**Pull request by cyliangtw**
### Description

This PR could support unhashable list of Mbed target.json .
Most list in target.json are string type and could be a set type of python.
However, some targets add integer list, like as:       
```
"sectors": [
            [
                0,
                2048
            ]

```

This kind type will encounter `TypeError: unhashable type: 'list'` in `source.py` for config preparation.

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

**Comment by Patater** [comment]
Hi @cyliangtw

Thanks a lot for your contirbution! We can confirm there is a bug here and that this fixes it. Unfortunately, we are unable to take in your change without unit test coverage. Could you please add testing for this bug, to ensure we don't accidentally break things again?

Thanks!

**Comment by cyliangtw** [comment]
Hi @Patater 

Did you mean to run tests folder - Unit and integration tests ?
Though I verified this modification OK by several targets of Mbed-OS in cmake build,
 I tried to run the tests of mbed-tools and refer to` [command]/bin/bash --noprofile --norc /home/vsts/work/_temp/3cafb457-bd5f-433a-bc63-d19b59973d2d.sh`.
However I can't see this shell file and got some usage error as below:
```
CYLiang@OANBE02780 MINGW32 /c/ARM_mbed/tools/mbed-tools (master)
$ pytest.exe ./tests/
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov --cov-report xml:coverage/coverage.xml --cov-report=html ./tests/
  inifile: C:\ARM_mbed\tools\mbed-tools\pytest.ini
  rootdir: C:\ARM_mbed\tools\mbed-tools

```
How to run these tests or any alternative way for unit test coverage ?

**Comment by ladislas** [comment]
@cyliangtw have you read this?

https://github.com/ARMmbed/mbed-tools/blob/master/CONTRIBUTING.md#development-and-testing

**Comment by cyliangtw** [comment]
Hi @ladislas ,
Thanks and I could run unit test.

**Comment by cyliangtw** [comment]
Hi @Patater ,
   In Run static analysis, I added commit c8b450c to fix  issue  `E111 indentation is not a multiple of four`.
About Run unit tests, it's fine and I attached the test log for reference.

```
tests\targets\_internal\targets_json_parsers\test_overriding_attribute_parser.py . [ 98%]
.........                                                                [100%]

----------- coverage: platform win32, python 3.7.8-final-0 -----------
Coverage HTML written to dir htmlcov
Coverage XML written to file coverage/coverage.xml


======================= 579 passed, 7 skipped in 13.21s =======================
```
[test.log](https://github.com/ARMmbed/mbed-tools/files/5925979/test.log)

**Comment by Patater** [comment]
Hi @cyliangtw

Could you please add a test to this file https://github.com/ARMmbed/mbed-tools/blob/master/tests/build/_internal/config/test_source.py? Please make sure the test fails when you don't have your fix in place, and that the test passes when you add your fix.

This will fix the CI issue on this PR from codecov, where there are lines being added that aren't covered by any tests.

```
src/mbed_tools/build/_internal/config/source.py#L49-L50

Added lines #L49 - L50 were not covered by tests
```

Thanks!

**Comment by cyliangtw** [comment]
> Hi @cyliangtw
> 
> Could you please add a test to this file https://github.com/ARMmbed/mbed-tools/blob/master/tests/build/_internal/config/test_source.py? Please make sure the test fails when you don't have your fix in place, and that the test passes when you add your fix.
> 
> This will fix the CI issue on this PR from codecov, where there are lines being added that aren't covered by any tests.
> 
> ```
> src/mbed_tools/build/_internal/config/source.py#L49-L50
> 
> Added lines #L49 - L50 were not covered by tests
> ```
> 
> Thanks!

Hi  @Patater ,
Yes, I added one more test of test_source.py to cover  source.py#L49-L50 in commit 2927a37.

```
Coverage for src\mbed_tools\build\_internal\config\source.py : 100% Show keyboard shortcuts
86 statements   
```

```
tests\targets\_internal\targets_json_parsers\test_overriding_attribute_parser.py . [ 98%]
.........                                                                [100%]

----------- coverage: platform win32, python 3.7.8-final-0 -----------
Coverage HTML written to dir htmlcov
Coverage XML written to file coverage/coverage.xml


======================= 579 passed, 7 skipped in 12.63s =======================
(.venv) [cygwin~mbed-tools]$

```

**Comment by Patater** [comment]
Hi @cyliangtw

Thanks for adding a test. Could you please rebase on the latest `master` branch? Please also ensure you pass linting locally with `tox -e linting`. Check for any files that may be missing newlines at the end and ensure they have newlines; GitHub will show these kinds of lines as a circle with a slash through it, like on a "No Smoking" sign.

Thanks!

**Comment by cyliangtw** [comment]
Hi @Patater ,
I rebase the mater branch and verified newlines checking OK by `tox -e linting`.
About `Analyse and Test Test Assert_news`, it seems relative to inherit rrequirements-test.txt issue in `[testenv:checknews]` test.
Temporarily mark rrequirements-test.txt of `[testenv]` in tox.ini, it could pass `tox -e checknews` as:
```
(.venv) [cygwin~mbed-tools]$ tox -e checknews
.package create: C:\ARM_mbed\tools\mbed-tools\.tox\.package
.package installdeps: setuptools>=42, wheel, setuptools_scm[toml]>=3.4
checknews installed: click==7.1.2,incremental==17.5.0,Jinja2==2.11.3,MarkupSafe==1.1.1,toml==0.10.2,towncrier==19.2.0
checknews runtests: PYTHONHASHSEED='678'
checknews runtests: commands[0] | python -m towncrier.check
On trunk, or no diffs, so no newsfragment required.
___________________________________ summary ___________________________________
  checknews: commands succeeded
  congratulations :)

```

**Comment by Patater** [comment]
@cyliangtw I believe you can get through the last CI failure by adding a news file to this PR, as [documented in our contributing guide](https://github.com/ARMmbed/mbed-tools/blob/master/CONTRIBUTING.md#news-files). Thank you!

**Comment by cyliangtw** [comment]
Hi @Patater,
According to my modification, it should be no matter with the prompt error message.
My environment is Windows, I really can't realize why it detect os.sync() of flash.py and invoke error of pre-commit.EXE:
```
e https://pre-commit.com/#using-the-latest-version-for-a-repository for more details.
licenseheaders...........................................................Passed
black....................................................................Passed
flake8...................................................................Passed
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src\mbed_tools\build\flash.py:24:9: error: Module has no attribute "sync"; maybe "fsync"?
Found 1 error in 1 file (checked 94 source files)

ERROR: InvocationError for command 'C:\\ARM_mbed\\tools\\mbed-tools\\.tox\\linting\\Scripts\\pre-commit.EXE run --all-files' (exited with code 1)

```

**Comment by Patater** [comment]
@cyliangtw I'm not sure where the Windows issue is coming from. We know that prep-release won't work on Windows, but we'd expect everything else to work OK. We test on Windows via Azure Pipelines and [the tests there are working OK](https://dev.azure.com/mbed-tools/mbed-tools/_build/results?buildId=4315&view=logs&j=d9478f6e-1d7d-5129-41cf-c6f93bd2bd8f&t=a9161cdc-088e-5feb-3cd4-e7fa743604af).

You've passed everything up on the GitHub-integrated CI except the news check. You can verify you've done the newsfile locally with `tox -e checknews`.

**Comment by cyliangtw** [comment]
> @cyliangtw I'm not sure where the Windows issue is coming from. We know that prep-release won't work on Windows, but we'd expect everything else to work OK. We test on Windows via Azure Pipelines and [the tests there are working OK](https://dev.azure.com/mbed-tools/mbed-tools/_build/results?buildId=4315&view=logs&j=d9478f6e-1d7d-5129-41cf-c6f93bd2bd8f&t=a9161cdc-088e-5feb-3cd4-e7fa743604af).
> 
> You've passed everything up on the GitHub-integrated CI except the news check. You can verify you've done the newsfile locally with `tox -e checknews`.

Hi @Patater ,
it could work well with `tox -e checknews` locally as:
```
(.venv) [cygwin~mbed-tools]$ tox -e checknews
.package create: C:\ARM_mbed\tools\mbed-tools\.tox\.package
.package installdeps: setuptools>=42, wheel, setuptools_scm[toml]>=3.4
checknews installed: click==7.1.2,incremental==17.5.0,Jinja2==2.11.3,MarkupSafe==1.1.1,toml==0.10.2,towncrier==19.2.0
checknews runtests: PYTHONHASHSEED='580'
checknews runtests: commands[0] | python -m towncrier.check
On trunk, or no diffs, so no newsfragment required.
___________________________________ summary ___________________________________
  checknews: commands succeeded
  congratulations :)

```

**Comment by Patater** [comment]
@cyliangtw Could you please ensure you add a news file to this PR? This PR should be ready to merge after that.

**Comment by ladislas** [comment]
@cyliangtw regarding the `news` file, I was also confused at first when doing my first PR.

I found those info really useful:

**Comment by ladislas** [comment]
@cyliangtw regarding the `news` file, I was also confused at first when doing my first PR.

I found those info really useful:

https://github.com/ARMmbed/mbed-tools/blob/master/CONTRIBUTING.md#news-files

Your `news` file should be added here, in tine `./news` directory:

https://github.com/ARMmbed/mbed-tools/tree/master/news

**Comment by cyliangtw** [comment]
Hi @Patater , @ladislas ,
Thanks of your suggestion, it could pass server `Analyse and Test Test Assert_news` after add news file in commit 850a74e.

**Comment by Patater** [comment]
I've updated the commit structure and commit messages. After CI is good, we can merge.

@cyliangtw Thanks a lot for finding this bug and raising a PR to fix it! Much appreciated.
@ladislas Thanks for your help on the review!

**Review by Patater — APPROVED** [review]

**Pull request merged**
